apps/server/src/modules/authentication/uc/account.uc.ts

```diff
@@ -15,45 +20,125 @@ export class AccountUc {
-			return false;
+		// TODO Check with BC-1471 - This might be removed from the account page
+		if (params.language && user.language !== params.language) {
+			// TODO Check if there is an enum or some other fixed values here?
+			user.language = params.language;
+			updateUser = true;
+		}
+		if (params.email && user.email !== params.email) {
+			const newMail = params.email.toLowerCase();
+			await this.checkUniqueEmail(account, user, newMail);
+			user.email = newMail;
+			account.username = newMail;
+			updateUser = true;
+			updateAccount = true;
 		}
 
-		const permissionsToCheck: string[] = [];
-		if (this.hasRole(targetUser, 'student')) {
-			permissionsToCheck.push('STUDENT_EDIT');
+		if (params.firstName && user.firstName !== params.firstName) {
+			if (!this.hasPermissionsToChangeOwnName(user)) {
+				throw new InvalidOperationError('No permission to change first name');
+			}
+			user.firstName = params.firstName;
+			updateUser = true;
 		}
-		if (this.hasRole(targetUser, 'teacher')) {
-			permissionsToCheck.push('TEACHER_EDIT');
+		if (params.lastName && user.lastName !== params.lastName) {
+			if (!this.hasPermissionsToChangeOwnName(user)) {
+				throw new InvalidOperationError('No permission to change last name');
+			}
+			user.lastName = params.lastName;
+			updateUser = true;
 		}
-		if (permissionsToCheck.length === 0) {
-			// target user is neither student nor teacher. Undefined what to do
-			return false;
+
+		if (updateUser) {
+			try {
+				await this.userRepo.update(user);
+			} catch (err) {
+				throw new EntityNotFoundError(User.name);
+			}
+		}
+		if (updateAccount) {
+			try {
+				await this.accountRepo.update(account);
+			} catch (err) {
+				throw new EntityNotFoundError(Account.name);
+			}
 		}
+	}
 
-		return this.permissionService.hasUserAllSchoolPermissions(currentUser, permissionsToCheck);
+	async changeMyTemporaryPassword(userId: EntityId, password: string, confirmPassword: string): Promise<void> {
+		if (password !== confirmPassword) {
+			throw new InvalidOperationError('Password and confirm password do not match.');
+		}
+
+		let user: User;
+		try {
+			user = await this.userRepo.findById(userId);
+		} catch (err) {
+			throw new EntityNotFoundError(User.name);
+		}
+		const userPreferences = <UserPreferences>user.preferences;
+
+		if (!user.forcePasswordChange && userPreferences.firstLogin) {
+			throw new InvalidOperationError('The password is not temporary, hence can not be changed.');
+		} // Password change was forces or this is a first logon for the user
+
+		let targetAccount: Account;
+		try {
+			targetAccount = await this.accountRepo.findByUserId(userId);
+		} catch (err) {
+			throw new EntityNotFoundError(Account.name);
+		}
+
+		if (targetAccount.system) {
+			throw new InvalidOperationError('External account details can not be changed.');
+		}
+
+		if (targetAccount.password === undefined || (await this.checkPassword(password, targetAccount.password))) {
```

Comment:
This means, if the old password is undefined, you can't set it.
I think you meant `targetAccount.password != null || (await this.checkPassword(password, targetAccount.password))`
(`!= null` checks that it is neither null nor undefined)

Reply:
Yeah, also I split the condition and check if there is a password anyway. If not that is an Error (inconsistent state). Otherwise we compare with previous and if those match this is a business error.